### PR TITLE
WIP: feat(#185):Add directory structure for feedback microservice

### PR DESCRIPTION
**Pull request by anjneeksharma**
This is part of feedback microservice that create directory structure for further development.  

## Does this PR introduce a breaking change?

 No

## Screenshots:

<details>
<summary>View Screenshots</summary>

image

</details>

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [x] Does the change have appropriate unit tests?
- [x] Did tests pass?
- [x] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [ ] Was this feature demo'd and the design review approved?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -49,5 +49,10 @@
     "eslint-plugin-promise": "4.2.1",
     "eslint-plugin-standard": "4.0.1",
     "lerna": "3.20.2"
+  },
```
Comment (riginoommen) [inline package.json:52]: Please checkout this file changes. also checkout package-lock.json

File: package-lock.json [hunk 2]
```diff
@@ -5,25 +5,36 @@
...
+      "resolved": "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.8.3.tgz",
+      "integrity": "sha512-a9gxpmdXtZEInkCSHUJDLHZVBgb1QS0jhss4cPP93EW7s+uC5bikET2twEF3KV+7rDblJcmNvTR7VJejqd2C2g==",
       "requires": {
-        "@babel/highlight": "^7.0.0"
+        "@babel/highlight": "^7.8.3"
       }
     },
+    "@babel/helper-validator-identifier": {
+      "version": "7.9.0",
+      "resolved": "https://registry.npmjs.org/@babel/helper-validator-identifier/-/helper-validator-identifier-7.9.0.tgz",
```
Comment (riginoommen) [inline package-lock.json:17]: Please checkout this file changes

File: packages/feedback-service/package.json [hunk 3]
```diff
@@ -0,0 +1,92 @@
+{
+  "name": "feedback",
```
Comment (riginoommen) [inline packages/feedback-service/package.json:2]: It should be feedback

Comment (anjneeksharma) [reply packages/feedback-service/package.json:2]: yes i can see that is feedback only ? please elaborate your comment.

File: packages/feedback-service/package.json [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  "scripts": {
+    "test": "jest --forceExit",
+    "build:dev": "webpack --config webpack.dev.js",
+    "start:dev": "webpack --watch --config webpack.dev.js",
+    "build": "webpack --config webpack.prod.js",
+    "start": "webpack --watch --config webpack.prod.js"
+  },
+  "author": {
+    "name": "Anjnee K Sharma",
+    "email": "anjnee.k.sharma@gmail.com"
```
Comment (deshmukhmayur) [inline packages/feedback-service/package.json:34]: @sayak-sarkar Shouldn't this be red hat email?

File: packages/feedback-service/service.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    const dbCredentials = (process.env.DB_USER && process.env.DB_PASSWORD)
+      ? `${process.env.DB_USER}:${process.env.DB_PASSWORD}@`
+      : '';
+    const dbConnection = `mongodb://${dbCredentials}${process.env.DB_PATH}/${process.env.DB_NAME}`;
+
+    mongoose.connect(dbConnection, { useNewUrlParser: true, useCreateIndex: true }).catch(console.error);
+
+    mongoose.connection.on('error', error => {
+      console.error(error);
+    });
```
Comment (deshmukhmayur) [inline packages/feedback-service/service.ts:39]: Decrease indent

File: packages/feedback-service/src/config/config.ts [hunk 6]
```diff
@@ -0,0 +1,538 @@
+const nodemailer = require("nodemailer");
```
Comment (deshmukhmayur) [inline packages/feedback-service/src/config/config.ts:1]: Why is this file here? Remove this file immediately. This has all the credentials and secrets. Don't just blindly copy anything.

Comment (deshmukhmayur) [reply packages/feedback-service/src/config/config.ts:1]: And also remove it from the commit history.

Comment (sayak-sarkar) [reply packages/feedback-service/src/config/config.ts:1]: @anjneeksharma Remove this file ASAP! This is a security issue. This is a public repository!

File: packages/feedback-service/src/types.d.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+declare module '*.graphql';declare module '*.json';
```
Comment (diwanshi) [inline packages/feedback-service/src/types.d.ts:1]: there should be a line break on this line

File: packages/feedback-service/.env.example [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+DB_NAME=db_name
+# DB_USER=
+# DB_PASSWORD=
+
+## Certificate Configuration
+SSL_KEY=/etc/pki/tls/private/server.key
+SSL_CERT=/etc/pki/tls/certs/server.crt
+
+## Cookie Parser
+COOKIE_PARSER=FEEDBACK_SECRET
```
[the remark this answers is not in the document]
Comment (diwanshi) [reply packages/feedback-service/.env.example:22]: @anjneeksharma FYI, when u create a service using generator, you need to copy contents of the example file and put it in.env file thats created on ur local machine.you are supposed to add all these values in ur local .env file and not commit it.

File: packages/feedback-service/package.json (hunk 3 above)
Comment (diwanshi) [reply packages/feedback-service/package.json:2]: @anjneeksharma he meant the `feedback` should be `Feedback` notice the caps

File: packages/feedback-service/service.ts (hunk 5 above)
Comment (diwanshi) [reply packages/feedback-service/service.ts:39]: @anjneeksharma I'm wondering how your indentation is changing from what we have in the generator?

File: packages/feedback-service/src/config/config.ts (hunk 6 above)
Comment (diwanshi) [reply packages/feedback-service/src/config/config.ts]: @anjneeksharma can you please remove this file from repo and commit history asap. Noone should be seeing this file.

File: packages/feedback-service/src/e2e/feedback.spec.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+        expect(res.body.data).toHaveProperty('delete');
+        expect(res.body.data.delete).toMatchObject({message: 'DELETE API for Feedback microservice'});
+      })
+      .end((err, res) => {
+        done(err);
+      });
+  });
+});
+
```
Comment (diwanshi) [inline packages/feedback-service/src/e2e/feedback.spec.ts:150]: update your tests

File: packages/feedback-service/src/e2e/mock.json [hunk 10]
```diff
@@ -0,0 +1,4 @@
+{
+   "message": "Define your test mocks in this file and remove this key value pair if not required"
```
Comment (diwanshi) [inline packages/feedback-service/src/e2e/mock.json:2]: update your mocks

File: packages/feedback-service/src/resolver.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
+import { Feedback } from './schema';
+import { pubsub } from './helpers';
+import { getTransporterInstance, getEnvironmentPaths  } from "./helpers";
+
```
Comment (diwanshi) [inline packages/feedback-service/src/resolver.ts:4]: remove unnecessary line breaks

File: packages/feedback-service/src/resolver.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    addFeedback(root: any, args: any, ctx: any) {
+      const data = new Feedback(args.input);
+      return data.save().then(response => {
+        Feedback.findById(response._id)
+          .populate("module")
+          .then(fb => {
+            if (fb) {
+              const feedbackfor = fb.portalFeedback ? "One Platfrom" : Object(fb.module).name + " module";
+              const feedbackType = fb.feedbackType === "Bug" ? "submitting a bug report" : "your feedback";
+              let body, imsg, summary;
```
Comment (diwanshi) [inline packages/feedback-service/src/resolver.ts:44]: can you explain what imsg means? or probably rewrite variable names to something meaningful and that are easy to understand

File: packages/feedback-service/src/resolver.ts [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+              } else if (fb.feedbackType === "Bug") {
+                // tslint:disable-next-line:max-line-length
+                body = `Hi ${fb.timestamp.createdBy.name},\n\nThanks for submitting your bug report on ${feedbackfor} with us.\n\nPlease find the details of the feedback we recieved as below:\nTitle: ${fb.title} ${summary}\nThe bug is currently under review and we will be looking into it shortly.\n\nIn the meanwhile, you can also read the feedback shared by others here: https://one.redhat.com/feedback \n\n${imsg}\n\nRegards,\nOne Portal Team \n\nP.S.: This is an automated email. Please do not reply.`;
+              }
+
+              const environmentPaths: any = getEnvironmentPaths();
```
Comment (diwanshi) [inline packages/feedback-service/src/resolver.ts:63]: environmentPaths should be defined at the top (commonly) so that u can use it for other mutations, queries or subscriptions

File: packages/feedback-service/src/schema.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
+import { Document, Model, model, Schema } from 'mongoose';
+import { IFeedback } from "./types";
+
+export const FeedbackSchema: Schema = new Schema({
+  description: String,
+  experience: String,
+  iid: String,
```
Comment (diwanshi) [inline packages/feedback-service/src/schema.ts:7]: what is iid?

File: packages/feedback-service/src/schema.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+  module: { type: Schema.Types.ObjectId, ref: "Module" },
+  portalFeedback: Boolean,
+  timestamp: {
```
Comment (diwanshi) [inline packages/feedback-service/src/schema.ts:10]: @deshmukhmayur i think we decided to not define timestamps like this in one platform?
can u confirm?

File: packages/feedback-service/src/resolver.ts (hunk 13 above)
Comment (deshmukhmayur) [reply packages/feedback-service/src/resolver.ts:63]: Actually this is being handled by the `.env` config.
So this entire environmentPaths can be removed.

File: packages/feedback-service/src/schema.ts (hunk 15 above)
Comment (deshmukhmayur) [reply packages/feedback-service/src/schema.ts:10]: Yes. AIt should have a flat structure.
So `createdAt` `modifiedAt` should be a direct child of the schema.

**Comment by riginoommen** [comment]
After reviewing the code with @anjneeksharma we found out that this branch is missing the following things

- Test Cases 
- Microservices has n't been tested in the development environment by executing `npm run build:dev && npm run start:dev` command


I suggest to close this PR and delete the branch from your fork also

**Comment by riginoommen** [comment]
s

**Comment by riginoommen** [comment]
reopening

File: packages/feedback-service/package.json (hunk 4 above)
Comment (sayak-sarkar) [reply packages/feedback-service/package.json:34]: Yup!
